Components: Introduce a basic ProgressBar component (#53030)
WordPress/gutenberg: packages/components/src/progress-bar/README.md

# ProgressBar

<div class="callout callout-alert">
This feature is still experimental. “Experimental” means this is an early implementation subject to drastic and breaking changes.
</div>

A simple horizontal progress bar component.

Supports two modes: determinate and indeterminate. A progress bar is determinate when a specific progress value has been specified (from 0 to 100), and indeterminate when a value hasn't been specified.

### Props

The component accepts the following props:

#### `value`: `number`

The progress value, a number from 0 to 100.
If a `value` is not specified, the progress bar will be considered indeterminate.

-   Required: No

##### `className`: `string`

A CSS class to apply to the underlying `div` element, serving as a progress bar track.

- Required: No

##### `id`: `string`

> The HTML `id` of the `progress` element. This is necessary to be able to accessibly associate the label with that element.

Reviewer: ```suggestion
Specify a custom HTML `id` for the `progress` element. This is necessary when you want to accessibly associate a custom `label` with that element.
```

I think we can be a bit more clear that this is optional, and only for specific custom use cases.

On the other hand, I [checked](https://codesandbox.io/s/nice-https-f22hyt?file=/src/App.js) what the labeling priority is when there are competing labels, and it seems to be `aria-labelledby` > `aria-label` > `<label for="my-id">`. Therefore, if a consumer wants to do a `<label for="my-progress-bar">`, they're also going to have to remove the default `aria-label` prop, which is non-obvious 🤔 It would be more straightforward to use `aria-labelledby` or `aria-label` if they want to override the label. In this case, we probably shouldn't be calling out the `id` prop explicitly in our docs, as it'd be highlighting a non-recommended approach? (It will just be silently forwarded as part of the rest props.)

Author: Since I removed the `id` completely and it's now part of the props we pass down to the underlying `progress` element, this is no longer relevant.

# ProgressBar

> A simple horizontal progress bar component.

Reviewer: ```suggestion
<div class="callout callout-alert">
This feature is still experimental. “Experimental” means this is an early implementation subject to drastic and breaking changes.
</div>

A simple horizontal progress bar component.
```

Supports two modes: determinate and indeterminate. A progress bar is determinate when a specific progress value has been specified (from 0 to 100), and indeterminate when a value hasn't been specified.

### Usage

```jsx
/**
 * WordPress dependencies
 */

> import { __experimentalProgressBar as ProgressBar } from '@wordpress/components';

Reviewer: Hmm, I'm wondering about whether we should really be showing these in the code snippets if it involves an `unlock()` 🤔 It's almost like we're endorsing it.

[`DropdownMenuV2`](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--default) is the first of our experimental components involving a lock/unlock, and for that one:

- The component is shown in the Storybook, but without a `import` code snippet.
- The component readme is not shown in the public docs (i.e. not in the manifest.json).

I don't remember having really discussed this with Marco, but maybe we should stick with the same pattern for now? I'll discuss it with the team, but we soon might need to add a new Storybook grouping like "Components (Private)" or something so we can provide `unlock()` instructions for Gutenberg devs, and also avoid confusion for third-party devs.

Author: Good point. I've removed the example that contains unlocking, removed the README from the public docs, and verified that the import is not present in Storybook.

export const MyProgressBar = () => {
	return <ProgressBar value={ 30 } />;
};
```

### Props

The component accepts the following props:

#### `value`: `number`

The progress value, a number from 0 to 100.
If a `value` is not specified, the progress bar will be considered indeterminate.

-   Required: No

#### `trackColor`: `string`

Optional color of the progress bar track.

-   Required: No

#### `indicatorColor`: `string`

> Optional color of the progress bar indicator.

Reviewer: I think this is a good use case for leveraging the new [`Theme`](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-theme--default) system rather than adding props to this component. Here is a [list of the CSS variables](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/theme/README.md#writing-themeable-components) that can be used to achieve that.

Author: Good call. Removed those props in e00dc4f161c8763fa846b857f4a7152f67c83d42.

I'm going to try out the theme system in #53032. 

Thanks 🙌
